File: config/claude/hooks/spec-kit-plannotator.py
#!/usr/bin/env python3
"""Run Plannotator on specs markdown files after Claude edits them."""

from __future__ import annotations

import json
import shutil
import os
import subprocess
import sys
from pathlib import Path


def _resolve_file_path(cwd: str, tool_input: dict[str, object]) -> Path | None:
    file_path = tool_input.get("file_path")
    if not isinstance(file_path, str) or not file_path:
        return None

    path = Path(file_path)
    if path.is_absolute():
        return path
    if not cwd:
        return None
    return Path(cwd) / path


def _is_specs_markdown_file(project_dir: Path, file_path: Path) -> bool:
    try:
        relative_path = file_path.resolve().relative_to(project_dir.resolve())
    except ValueError:
        return False

    return relative_path.parts[:1] == ("specs",) and relative_path.suffix.lower() == ".md"


def _print_context(message: str, *, block: bool = False) -> None:
    payload: dict[str, object] = {}
    if block:
        payload["decision"] = "block"
        payload["reason"] = message
    else:
        payload["hookSpecificOutput"] = {
            "hookEventName": "PostToolUse",
            "additionalContext": message,
        }

    print(json.dumps(payload))


def main() -> None:
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError as exc:
        print(f"Error: Invalid JSON input: {exc}", file=sys.stderr)
        sys.exit(1)

    cwd = input_data.get("cwd", "")
    tool_input = input_data.get("tool_input", {})
    if not isinstance(tool_input, dict):
        sys.exit(0)

    file_path = _resolve_file_path(cwd, tool_input)
    if file_path is None:
        sys.exit(0)

    project_dir = Path(cwd) if cwd else file_path.parent
    if not _is_specs_markdown_file(project_dir, file_path):
        sys.exit(0)

    plannotator = shutil.which("plannotator")
    if not plannotator:
        _print_context(
            f"Plannotator was not run for `{file_path}` because the `plannotator` executable is not installed."
        )
        sys.exit(0)

    result = subprocess.run(
        [plannotator, "annotate", str(file_path), "--hook"],
        capture_output=True,
        text=True,
        env={**os.environ, "NO_COLOR": "1"},
    )

    stdout = result.stdout.strip()
    stderr = result.stderr.strip()

    if result.returncode != 0:
        message = (
            f"Plannotator failed for `{file_path}` with exit code {result.returncode}."
        )
        if stderr:
            message = f"{message}\n\n{stderr}"
        _print_context(message, block=True)
        sys.exit(0)

    if not stdout:
        sys.exit(0)

    try:
        payload = json.loads(stdout)
    except json.JSONDecodeError:
        _print_context(
            f"Plannotator returned non-JSON hook output for `{file_path}`.\n\n{stdout}",
            block=True,
        )
        sys.exit(0)

    print(json.dumps(payload))


if __name__ == "__main__":
    main()

File: config/opencode/plugins/spec-kit-plannotator.ts
import type { Plugin } from "@opencode-ai/plugin"
import path from "node:path"

const COOLDOWN_MS = 5000

const lastTriggeredAt = new Map<string, number>()
const activeProcesses = new Map<string, Promise<void>>()

function toAbsolutePath(projectDir: string, file: string): string {
  return path.isAbsolute(file) ? file : path.join(projectDir, file)
}

type HookResult = {
  decision?: "allow" | "block"
  reason?: string
}

function isMarkdownFile(projectDir: string, file: string): boolean {
  const absolutePath = toAbsolutePath(projectDir, file)
  const relativePath = path.relative(projectDir, absolutePath)

  if (relativePath.startsWith("..") || path.isAbsolute(relativePath)) {
    return false
  }

  return path.extname(relativePath).toLowerCase() === ".md"
}

function parseHookResult(stdout: string): HookResult | undefined {
  try {
    const parsed = JSON.parse(stdout) as HookResult
    if (parsed.decision === "allow" || parsed.decision === "block") {
      return parsed
    }
  } catch {
    return undefined
  }

  return undefined
}

function shouldTrigger(file: string): boolean {
  const now = Date.now()
  const lastRun = lastTriggeredAt.get(file) ?? 0

  if (now - lastRun < COOLDOWN_MS) {
    return false
  }

  lastTriggeredAt.set(file, now)
  return true
}

function getEventSessionID(event: {
  properties?: Record<string, unknown>
  sessionID?: unknown
  sessionId?: unknown
}): string | undefined {
  const candidates = [
    event.properties?.sessionID,
    event.properties?.sessionId,
    event.sessionID,
    event.sessionId,
  ]

  return candidates.find(
    (value): value is string => typeof value === "string" && value.length > 0,
  )
}

async function readProcessStream(
  stream: ReadableStream<Uint8Array> | null | undefined,
): Promise<string> {
  if (!stream) {
    return ""
  }

  return (await new Response(stream).text()).trim()
}

export const SpecKitPlannotatorPlugin: Plugin = async ({ client, directory }) => {
  let lastSessionID: string | undefined

  async function launchAnnotator(file: string, sessionID?: string) {
    const absolutePath = toAbsolutePath(directory, file)

    if (activeProcesses.has(absolutePath) || !shouldTrigger(absolutePath)) {
      return
    }

    const plannotator = Bun.which("plannotator")
    if (!plannotator) {
      await client.app.log({
        body: {
          service: "spec-kit-plannotator",
          level: "warn",
          message: "Skipping markdown annotation because `plannotator` is not installed",
          extra: { file: absolutePath },
        },
      })
      return
    }

    const proc = Bun.spawn([plannotator, "annotate", absolutePath, "--hook"], {
      stdin: "ignore",
      stdout: "pipe",
      stderr: "pipe",
      env: { ...globalThis.process.env, NO_COLOR: "1" },
    })

    const run = (async () => {
      const stdoutPromise = readProcessStream(proc.stdout)
      const stderrPromise = readProcessStream(proc.stderr)
      const exitCode = await proc.exited
      const [stdout, stderr] = await Promise.all([stdoutPromise, stderrPromise])

      if (exitCode !== 0) {
        await client.app.log({
          body: {
            service: "spec-kit-plannotator",
            level: "error",
            message: "Plannotator annotate command failed",
            extra: {
              file: absolutePath,
              exitCode,
              stderr,
              stdout,
            },
          },
        })
        return
      }

      if (!stdout) {
        return
      }

      const hookResult = parseHookResult(stdout)
      if (!hookResult) {
        await client.app.log({
          body: {
            service: "spec-kit-plannotator",
            level: "error",
            message: "Plannotator returned non-JSON hook output",
            extra: { file: absolutePath, stdout },
          },
        })
        return
      }

      if (hookResult.decision !== "block" || !hookResult.reason) {
        return
      }

      if (!sessionID) {
        await client.app.log({
          body: {
            service: "spec-kit-plannotator",
            level: "warn",
            message:
              "Received Plannotator annotations but could not route them back to OpenCode because no session ID was available",
            extra: { file: absolutePath },
          },
        })
        return
      }

      try {
        await client.session.prompt({
          path: { id: sessionID },
          body: {
            parts: [
              {
                type: "text",
                text: hookResult.reason,
              },
            ],
          },
        })
      } catch (error) {
        await client.app.log({
          body: {
            service: "spec-kit-plannotator",
            level: "error",
            message: "Failed to send Plannotator annotations back to OpenCode",
            extra: {
              file: absolutePath,
              sessionID,
              error: error instanceof Error ? error.message : String(error),
            },
          },
        })
      }
    })()

    activeProcesses.set(absolutePath, run)

    await client.app.log({
      body: {
        service: "spec-kit-plannotator",
        level: "info",
        message: "Opening Plannotator for markdown file",
        extra: { file: absolutePath, sessionID },
      },
    })

    void run.finally(() => {
      activeProcesses.delete(absolutePath)
    })
  }

  return {
    event: async ({ event }) => {
      const sessionID = getEventSessionID(event)
      if (sessionID) {
        lastSessionID = sessionID
      }

      const targetSessionID = sessionID ?? lastSessionID

      if (event.type === "file.edited") {
        if (isMarkdownFile(directory, event.properties.file)) {
          await launchAnnotator(event.properties.file, targetSessionID)
        }
        return
      }

      if (event.type === "file.watcher.updated") {
        if (event.properties.event === "unlink") {
          return
        }

        if (isMarkdownFile(directory, event.properties.file)) {
          await launchAnnotator(event.properties.file, targetSessionID)
        }
      }
    },
  }
}

export default SpecKitPlannotatorPlugin
